### DISCUSSION TITLE
`AddSystemFonts` overloads

### DISCUSSION THREAD

**Discussion by LeaFrock** [SixLabors/Fonts#239] [Ideas] [answered]
I notice that `FontCollectionExtensions` provides `AddSystemFonts` method.

Is it possible to add a overload method like below? This method can filter the fonts not wanted.

```csharp
        public static FontCollection AddSystemFonts(this FontCollection collection, Predicate<FontMetrics> predicate)
        {
            foreach (FontMetrics? metric in (IReadOnlyFontMetricsCollection)SystemFonts.Collection)
            {
                if(predicate(metric))
                {
                    ((IFontMetricsCollection)collection).AddMetrics(metric);
                }
            }

            return collection;
        }
```

**Comment by tocsoft** [comment, marked as answer]
@LeaFrock can you provide use case(s) why this is needed? How the current API is causing issues etc. it'll help us priorities API changes etc.

**Reply by LeaFrock** [reply]
For example, if I want to generate captcha images with system fonts(one character to one font), I don't need all of them. I want to filter the ones which `CultureInfo` is not `en-us`, or `Name` not start with `Microsoft`. It depends on requirements of development.

The original thought comes from the problem of #237 . After upgrading to the lastest version of `ImageSharp.Fonts` and trying to use `AddSystemFonts`, I've met this problem when some of special fonts is randomed for a char. Therefore, this overload method can also be a temp solution before fixing similar problems.

**Reply by JimBobSquarePants** [reply]
>I've met this problem when some of special fonts is randomed for a char

Can you please clarify what you mean by this as it doesn't make any sense to me.

**Reply by LeaFrock** [reply]
@JimBobSquarePants  I want to generate a captcha image, and each of the chars has a random font from system fonts. 

### Examples(Win10)
#### Good case
![1645066874(1)](https://user-images.githubusercontent.com/20787265/154396560-190d0340-62c9-443f-9686-b0cc4ea3dda5.png)

The 4 random fonts are the following,

```
DengXian
DengXian
Calibri Light
MS UI Gothic
```

That's fine.

#### Bad case 1

![1645066501(1)](https://user-images.githubusercontent.com/20787265/154395840-9bd5b4ca-0ed7-4285-97a4-b02ab5337cb8.png)

The 4 random fonts are the following,

```
Segoe MDL2 Assets
DengXian Light
Tahoma
Sitka Small
```

You can see that `Segoe MDL2 Assets` is not a available font for me.

#### Bad case 2

![1645067155(1)](https://user-images.githubusercontent.com/20787265/154397062-3127365e-c788-4424-8efb-12f3bdb4097f.png)

It happens when using this font `Microsoft Sans Serif`. I think it may be fixed by issue #237 .

### Conclusion 

That's why I hope to add an overload method which filters the original fonts not usable for my requirement.

**Reply by JimBobSquarePants** [reply]
@tocsoft Current table support issues aside this looks like a pretty useful extension that would add a negligible burden in additional maintenance . What do you think?

**Reply by LeaFrock** [reply]
Or just change the current method to a compatible one like below, 

``` csharp
        public static FontCollection AddSystemFonts(this FontCollection collection, Predicate<FontMetrics>? predicate = default)
        {
            if (predicate is null)
            {
                foreach (FontMetrics? metric in (IReadOnlyFontMetricsCollection)SystemFonts.Collection)
                {
                    ((IFontMetricsCollection)collection).AddMetrics(metric);
                }
            }
            else
            {
                foreach (FontMetrics? metric in (IReadOnlyFontMetricsCollection)SystemFonts.Collection)
                {
                    if (predicate(metric))
                    {
                        ((IFontMetricsCollection)collection).AddMetrics(metric);
                    }
                }
            }
            return collection;
        }
```

**Reply by tocsoft** [reply]
> @tocsoft Current table support issues aside this looks like a pretty useful extension that would add a negligible burden in additional maintenance . What do you think?

Yeah agreed, I'm personally not huge fan of doing anything that gives the appearance of encouraging the use of system fonts at all, as its just a never ending source of issues (that just filtering is a workaround for) that don't exist if you ship our own fonts with your code. 

However I am not opposed enough to try and block this API proposal.

The follow up question however is how breaking a change do we want... it will cause less headaches if we just add a new overload.

**Reply by JimBobSquarePants** [reply]
Since we try to avoid default values in methods a new overload seems sound. @LeaFrock Fancy raising a PR?

**Reply by LeaFrock** [reply]
> Fancy raising a PR?

With pleasure.
